src/wallet/wallet.cpp

```diff
@@ -81,13 +81,33 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
+static std::mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cond;
+
+void WaitForWalletRelease(std::shared_ptr<CWallet>& wallet)
+{
+    std::weak_ptr<CWallet> weak_wallet(wallet);
+    wallet.reset();
+
+    std::unique_lock<std::mutex> lk(g_wallet_release_mutex);
+    while (!weak_wallet.expired()) {
+        g_wallet_release_cond.wait(lk);
+    }
+}
+
 // Custom deleter for shared_ptr<CWallet>.
 static void ReleaseWallet(CWallet* wallet)
 {
-    wallet->WalletLogPrintf("Releasing wallet\n");
-    wallet->BlockUntilSyncedToCurrentChain();
-    wallet->Flush();
-    delete wallet;
+    {
+        std::unique_lock<std::mutex> lk(g_wallet_release_mutex);
+
+        wallet->WalletLogPrintf("Releasing wallet\n");
+        wallet->BlockUntilSyncedToCurrentChain();
+        wallet->Flush();
+        delete wallet;
+    }
+
+    g_wallet_release_cond.notify_all();
```

Comment:
I think we need to confirm that `!weak_wallet.expired()` in the other thread is true at the point of this call (and guaranteed by the c++ standard to be true). Otherwise there could be a lost-wakeup bug here because this would be signalling the condition variable before the condition is true.

Reply:
Good catch, I'll investigate.

Reply:
This no longer applies, @ryanofsky concern was valid.